### PR TITLE
New snapshot save function and multi-gpu tests

### PR DESCRIPTION
**Pull request by kan-bayashi**
Save trainer / model / optimizer states into one file in pytorch.

- Added `torch_snapshot` to save trainer / model / optimizer states into one file
- Added `torch_resume` to resume training from new snapshot
- Unified `torch_save` and `torch_load` functions in ASR and TTS
- Added multi-gpu tests for pytorch
- Added save / load test for pytorch
- Added more docstrings

### CODE REVIEW

File: src/asr/asr_utils.py [hunk 1]
```diff
@@ -235,6 +204,87 @@ def _plot_and_save_attention(self, att_w, filename):
...
+    return torch_snapshot
+
+
+def _torch_snapshot_object(trainer, target, filename, savefun):
+    # make snapshot_dict dictionary
+    s = DictionarySerializer()
+    s.save(trainer)
+    if hasattr(trainer.updater.model, "model"):
+        # (for TTS)
+        if hasattr(trainer.updater.model.model, "module"):
```
Comment (sw005320) [inline src/asr/asr_utils.py]: Why TTS has a bit redundant model structure (`model.model`)?
Is there a way to make it same as ASR?

Comment (kan-bayashi) [reply src/asr/asr_utils.py]: If we assume that same options are used when resuming, it will be more simple.

But if you want to keep same format as model.loss.best, current format is needed.

**Comment by kan-bayashi** [comment]
@sw005320 Made more simple. which is better?
Previous: may work if we change options when resuming, but complex
Now: not work if we change options but simple

**Comment by sw005320** [comment]
My preference is previous one.
The code becomes complex in `asr_utils.py`, but this is OK.

**Comment by kan-bayashi** [comment]
Reverted.

**Comment by kan-bayashi** [comment]
Updated.
- Replaced loading part in `asr_pytorch.recog` with `torch_load`
- Update `torch_load`
    - Both model file and snapshot file to be used in decoding phase
- Added some cleaning part (remove loaded object explicitly)

**Comment by kan-bayashi** [comment]
Added more test codes.

**Comment by kan-bayashi** [comment]
@sw005320 I want to merge. Is it OK?
